### PR TITLE
VIMC-4405: Tasks can timeout

### PR DESCRIPTION
**Pull request by r-ash**
Builds on #42 (should be checked after that)
This PR will
* Add a preroute hook which checks for any timed out tasks on the workers
* Sets the default to 3 hours for a task

### CODE REVIEW

File: R/runner.R [hunk 1]
```diff
@@ -292,10 +289,44 @@ orderly_runner_ <- R6::R6Class(
...
+      incomplete <- logs[logs$command == "TASK_START", ]
+      if (nrow(incomplete) == 0) {
+        return(invisible(NULL))
+      }
+      incomplete$timeout <- as.numeric(unlist(
+        self$con$HMGET(self$keys$task_id_timeout, incomplete$message)))
+      now <- sprintf("%.04f", Sys.time()) ## Use same format as rrq
+      to_kill <- incomplete[incomplete$time + incomplete$timeout < now, ]
+
+      killed <- vapply(seq_len(nrow(to_kill)), function(row) {
```
Comment (r-ash) [inline R/runner.R:314]: I don't love this - is there a nicer way off hand to organise this code?

File: tests/testthat/test-runner.R [hunk 2]
```diff
@@ -359,10 +359,87 @@ test_that("status missing ID", {
...
+  path <- orderly_prepare_orderly_example("interactive", testing = TRUE)
+  runner <- orderly_runner(path)
+
+  ## Setup mock queue with realistic worker log
+  key1 <- runner$submit_task_report("interactive", timeout = 0)
+  Sys.sleep(0.5) ## Sleep to wait for runner to pick up task
+  logs <- runner$queue$worker_log_tail()
+
+  queue <- runner$queue
+  mock_queue <- list(
```
Comment (r-ash) [inline tests/testthat/test-runner.R:427]: This is pretty horrible too. I want to create a test to cover the case where in between getting the logs and the queue saying that task A can be killed it then completing meaning that the call to cancel the task run fails. I could alternatively create a helper function for returning the logs and then mock that out but that feels like a unnecessary extra function, but perhaps is justified if it makes test writing easier. Any thoughts?

File: tests/testthat/test-runner.R [hunk 3]
```diff
@@ -359,10 +359,87 @@ test_that("status missing ID", {
...
+  testthat::skip_on_cran()
+  skip_on_appveyor()
+  skip_on_windows()
+  skip_if_no_redis()
+  path <- orderly_prepare_orderly_example("demo")
+  runner <- orderly_runner(path, workers = 2)
+
+  key1 <- runner$submit_task_report("minimal", timeout = 20)
+  key2 <- runner$submit_task_report("slow10", timeout = 20)
+  Sys.sleep(1) ## Sleep so both reports get started
```
Comment (richfitz) [inline tests/testthat/test-runner.R:395]: this looks potentially flakey, especially if minimal might complete within time.

in other contexts I've rigged up things to poll status to wait until both things are started, which is probably preferable to the sleep as CI systems vary in speed. But minimal can be completed quite quickly so might escape kill by finishing

File: tests/testthat/test-runner.R (hunk 2 above)
Comment (richfitz) [reply tests/testthat/test-runner.R:427]: this is a hard thing to do without exposing a lot of details

Can you add this comment as a code comment so we know why the test exists and if things work emperically and the design as settled down we can look at making these tests more robust as a separate job?

**Review by richfitz — CHANGES_REQUESTED** [review]
Some fairly small comments - I think there's stuff we could improve in the testing but this already represents a big improvement over the existing implementation so let's now sweat it now

File: R/runner.R [hunk 4]
```diff
@@ -213,30 +213,27 @@ orderly_runner_ <- R6::R6Class(
...
-      ## Add a check_timeout method as a hook on the api build which calls this
-      ## SO then old jobs get killed
-      self$con$HSET(self$keys$key_timeout, key, timeout)
       key_report_id <- self$keys$key_report_id
       task_id <- self$submit(quote(
         orderly.server:::runner_run(key_report_id, key, root, name,   # nolint
                                     parameters, instance, ref, poll = poll)))
       self$con$HSET(self$keys$key_task_id, key, task_id)
       self$con$HSET(self$keys$task_id_key, task_id, key)
+      self$con$HSET(self$keys$task_id_timeout, task_id, timeout)
```
[the remark this answers is not in the document]
Comment (r-ash) [reply R/runner.R:236]: I think there is no cleaning up of these at the moment, when should we do that? I could do that right at the end of `runner_run` or should I do that later? Could do it when result is retrieved or when status is called and it says it is complete? Status one sounds like a good idea

[the remark this answers is not in the document]
Comment (richfitz) [reply R/runner.R:236]: at the end when we store status complete sounds good to me

File: R/runner.R (hunk 1 above)
Comment (r-ash) [reply R/runner.R]: thanks yeah, that is neater!

Comment (r-ash) [reply R/runner.R]: Added a note about `to_kill[, "message"]` to the code - it holds the task_id

File: tests/testthat/test-runner.R (hunk 3 above)
Comment (r-ash) [reply tests/testthat/test-runner.R:395]: Thanks yeah, used a `wait_for_id` which will be less flaky. So will tell us that reports have started. Really this test is checking that a report running with a long timeout won't get killed to running the minimal report does nothing for us here. I will remove that to make test simpler

File: R/api.R [hunk 5]
```diff
@@ -307,3 +308,9 @@ endpoint_run_metadata <- function(runner) {
    returning = returning_json("RunMetadata.schema")
  )
 }
+
+check_timeout <- function(runner) {
+  function() {
```
[the remark this answers is not in the document]
Comment (r-ash) [reply R/api.R:313]: So taking around 1ms when nothing to kill but then 4-5ms when killing 1 task. Which will be longer with more workers. I've added a throttle copying stuff from hintr.

**Review by richfitz — APPROVED** [review]

**Pull request merged**